### PR TITLE
[Cosmos3] Add cosmos3 Reasoner to llm only inference

### PR DESCRIPTION
**Pull request by Zhylkaaa**
## Motivation

Cosmos3 Omni model has 2 transformer towers stacked on top of each other for video generation. First tower is understanding tower that is also trained for text/visual reasoning. Metrics are included in model card under Reasoning Benchmarks: https://huggingface.co/nvidia/Cosmos3-Nano#reasoning-benchmarks .
This PR adds support for Reasoner only inference.

## Modifications

Should be merged after: https://github.com/sgl-project/sglang/pull/31590

* Added new model type cosmos3_omni
* Modified DefaultModelLoader with allow_patterns_overrides argument to allow reading Diffusers format checkpoints
* Added Cosmos3 specific checkpoint key mappings
* Extended Qwen3VL execution path checks with new model type (Cosmos3 Reason follows Qwen3VL architecture)

## Accuracy Tests

Simple request to make sure model generates reasonable output, since rest of the implementation relies on Qwen3VL integration.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #33335794643](https://github.com/sgl-project/sglang/actions/runs/33335794643)
Latest PR Test (Extra): :x: [Run #33335794466](https://github.com/sgl-project/sglang/actions/runs/33335794466)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33335794575](https://github.com/sgl-project/sglang/actions/runs/33335794575)

### CODE REVIEW

File: python/sglang/srt/models/cosmos3.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        config: Cosmos3Config,
+        quant_config: Optional[QuantizationConfig] = None,
+        prefix: str = "",
+    ) -> None:
+        super().__init__(config=config, quant_config=quant_config, prefix=prefix)
+
+        # The vision encoder weights live in a separate ``vision_encoder/``
+        # subfolder, so load them as a secondary weight source.
+        server_args = get_global_server_args()
+        self.secondary_weights = [
```
Comment (mickqian) [inline python/sglang/srt/models/cosmos3.py]: [P2] `--language-model-only` still registers the vision source. The parent has no visual tower in this mode, so the loader downloads and deserializes `vision_encoder` weights only to skip them. Please add the source only when `not self.language_model_only`, as Cosmos3 Edge does.

File: python/sglang/multimodal_gen/configs/pipeline_configs/cosmos3.py [hunk 2]
```diff
@@ -134,6 +134,14 @@ def __post_init__(self):
         self.vae_config.use_parallel_encode = False
         self.vae_config.use_parallel_decode = True
 
+    def update_config_from_dict(self, args, prefix: str = "") -> None:
```
Comment (mickqian) [inline python/sglang/multimodal_gen/configs/pipeline_configs/cosmos3.py]: [P2] This is now the first of two `update_config_from_dict` definitions. The following definition replaces it, so this distilled-checkpoint branch never runs. Please merge this into the existing method or remove the duplicate.

**Review by mickqian — CHANGES_REQUESTED** [review]
Please address the inline comments before merge.

File: python/sglang/srt/models/cosmos3.py (hunk 1 above)
Comment (kediwu0331) [reply python/sglang/srt/models/cosmos3.py]: resolved

File: python/sglang/multimodal_gen/configs/pipeline_configs/cosmos3.py (hunk 2 above)
Comment (kediwu0331) [reply python/sglang/multimodal_gen/configs/pipeline_configs/cosmos3.py]: resolved thank you

**Comment by kediwu0331** [comment]
@mickqian , thanks these are resolved. Could you review again?
And regarding the cicd test failure, bot explained it was due to missing run-ci label for call-gate job. But I don't have access to update PR labels. Do you know what's the cause of that required test failure?

**Comment by kediwu0331** [comment]
@mickqian Could you help review this as well? Let me know if there is any further changes needed before merging.

File: test/registered/unit/models/test_cosmos3.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+        # Importing common runs the AutoConfig registration side effects.
+        from transformers import AutoConfig
+
+        import sglang.srt.utils.hf_transformers.common  # noqa: F401
+
+        cfg = AutoConfig.for_model(
+            "cosmos3_omni",
+            text_config={"hidden_size": 128, "num_hidden_layers": 2},
+        )
+        self.assertIsInstance(cfg, Cosmos3Config)
```
Comment (rwang5203) [inline test/registered/unit/models/test_cosmos3.py]: This assertion fails with pinned Transformers 5.12.1 because its native Cosmos3OmniConfig owns this model type, leaving the registered CPU suite red. Try to assert the required config contract instead of one implementation class.

File: test/registered/unit/models/test_cosmos3.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        safetensors.torch.save_file(
+            {"llm": torch.zeros(2)},
+            os.path.join(root, "transformer", "diffusion_pytorch_model.safetensors"),
+        )
+        safetensors.torch.save_file(
+            {"vit": torch.zeros(2)},
+            os.path.join(root, "vision_encoder", "model.safetensors"),
+        )
+
+    @patch("sglang.srt.model_loader.loader.get_global_server_args")
```
Comment (rwang5203) [inline test/registered/unit/models/test_cosmos3.py]: `get_global_server_args` is no longer defined at this module path, and the same patch target appears again at lines 276 and 288. As a result, all three loader tests fail during mock setup and never exercise the new allow-pattern behavior. Could you update all three tests to patch `get_server_args` and return `None`, or initialize the current runtime context instead?

**Comment by rwang5203** [comment]
Neither checkpoint appears in the multimodal language model table, so users cannot discover this LLM serving support or its modes. Could you add Nano and Edge with a reproducible launch and media request?

File: python/sglang/srt/models/cosmos3.py (hunk 1 above)
Comment (rwang5203) [reply python/sglang/srt/models/cosmos3.py]: The vision-source guard is fixed, however both architectures still fail `--language-model-only` at the server allowlist before model construction. Please add both and make `/model_info` report text-only capability, since media requests otherwise return 400.

File: docs/docs/supported-models/multimodal_language_models.mdx [hunk 5]
```diff
@@ -349,6 +361,45 @@ response = requests.post(url, json=data)
...
+      "content": [
+        {"type": "text", "text": "Describe this image."},
+        {"type": "image_url", "image_url": {"url": "https://raw.githubusercontent.com/sgl-project/sglang/main/examples/assets/example_image.png"}}
+      ]
+    }],
+    "max_tokens": 128
+  }'
+```
+
+To serve only the text model (skipping the vision tower entirely), add `--language-model-only`. In this mode `/model_info` reports no image or video understanding and media requests are rejected:
```
Comment (rwang5203) [inline docs/docs/supported-models/multimodal_language_models.mdx]: `/model_info` has no video capability field, so this wording promises a signal clients can't observe. Could we name `has_image_understanding` and separately say that image and video requests are rejected?

File: python/sglang/srt/models/cosmos3.py (hunk 1 above)
Comment (rwang5203) [reply python/sglang/srt/models/cosmos3.py]: Resolved and closed.

File: test/registered/unit/models/test_cosmos3.py (hunk 3 above)
Comment (rwang5203) [reply test/registered/unit/models/test_cosmos3.py]: Resolved and closed.

File: test/registered/unit/models/test_cosmos3.py (hunk 4 above)
Comment (rwang5203) [reply test/registered/unit/models/test_cosmos3.py]: Resolved and closed.

**Review by rwang5203 — APPROVED** [review]

File: docs/docs/supported-models/multimodal_language_models.mdx (hunk 5 above)
Comment (kediwu0331) [reply docs/docs/supported-models/multimodal_language_models.mdx]: resolved thank you

Comment (rwang5203) [reply docs/docs/supported-models/multimodal_language_models.mdx]: Resolved. Appreciate the prompt responses. I have already approved this, great work!

**Comment by kediwu0331** [comment]
@rwang5203 I fixed failed test cases related to cosmos. It seems the required test cases are passed now. Could you help merge? Thanks

**Comment by kediwu0331** [comment]
@mickqian Could you also help merge this PR? Thanks!

**Review by mickqian — APPROVED** [review]
